ddtrace/contrib/aiokafka/patch.py

```diff
@@ -0,0 +1,248 @@
+
+config._add(
+    "kafka",
+    dict(
+        _default_service=schematize_service_name("kafka"),
+        distributed_tracing_enabled=asbool(os.getenv("DD_KAFKA_PROPAGATION_ENABLED", default=False)),
+    ),
+)
+
+
+def get_version():
+    # type: () -> str
+    return getattr(aiokafka, "__version__", "")
+
+
+class TracedAIOKafkaProducerMixin:
+    def __init__(self, *args, **kwargs):
+        super(TracedAIOKafkaProducerMixin, self).__init__(*args, **kwargs)
+        self._dd_bootstrap_servers = kwargs.get("bootstrap_servers")
+
+
+class TracedAIOKafkaConsumerMixin:
+    def __init__(self, *args, **kwargs):
+        super(TracedAIOKafkaConsumerMixin, self).__init__(*args, **kwargs)
+
+
+class TracedAIOKafkaProducer(TracedAIOKafkaProducerMixin, aiokafka.AIOKafkaProducer):
+    pass
+
+
+class TracedAIOKafkaConsumer(TracedAIOKafkaConsumerMixin, aiokafka.AIOKafkaConsumer):
+    pass
+
+
+def patch():
+    if getattr(aiokafka, "_datadog_patch", False):
+        return
+    aiokafka._datadog_patch = True
+
+    aiokafka.AIOKafkaProducer = TracedAIOKafkaProducer
+    aiokafka.AIOKafkaConsumer = TracedAIOKafkaConsumer
+
+    trace_utils.wrap(TracedAIOKafkaProducer, "send", traced_send)
+    trace_utils.wrap(TracedAIOKafkaProducer, "send_and_wait", traced_send_and_wait)
+    trace_utils.wrap(TracedAIOKafkaConsumer, "getone", traced_getone)
+    trace_utils.wrap(TracedAIOKafkaConsumer, "getmany", traced_getmany)
+    trace_utils.wrap(TracedAIOKafkaConsumer, "commit", traced_commit)
+
+    Pin().onto(aiokafka.AIOKafkaProducer)
+    Pin().onto(aiokafka.AIOKafkaConsumer)
+
+
+def unpatch():
+    if getattr(aiokafka, "_datadog_patch", False):
+        aiokafka._datadog_patch = False
+
+    if trace_utils.iswrapped(TracedAIOKafkaProducer.send):
+        trace_utils.unwrap(TracedAIOKafkaProducer, "send")
+    if trace_utils.iswrapped(TracedAIOKafkaProducer.send_and_wait):
+        trace_utils.unwrap(TracedAIOKafkaProducer, "send_and_wait")
+    if trace_utils.iswrapped(TracedAIOKafkaConsumer.getone):
+        trace_utils.unwrap(TracedAIOKafkaConsumer, "getone")
+    if trace_utils.iswrapped(TracedAIOKafkaConsumer.getmany):
+        trace_utils.unwrap(TracedAIOKafkaConsumer, "getmany")
+    if trace_utils.iswrapped(TracedAIOKafkaConsumer.commit):
+        trace_utils.unwrap(TracedAIOKafkaConsumer, "commit")
+
+    aiokafka.AIOKafkaProducer = _AIOKafkaProducer
+    aiokafka.AIOKafkaConsumer = _AIOKafkaConsumer
+
+
+async def traced_send(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    topic = get_argument_value(args, kwargs, 0, "topic")
+    value = get_argument_value(args, kwargs, 1, "value", True) or None
+    key = get_argument_value(args, kwargs, 2, "key", True) or ""
+    partition = get_argument_value(args, kwargs, 3, "partition", True) or None
+    headers = get_argument_value(args, kwargs, 5, "headers", True) or []
+
+    with pin.tracer.trace(
```

Comment:
Code in contrib/ shouldn't directly call `trace`, instead using `core.context_with_data` paired with an event handler implemented in `_trace/trace_handlers`.

Reply:
Ok, so there's one event being fired by the wrapped functions. And that event gets dealt with by:
- `_trace/tracehandlers` for normal tracing functionality
- `internal/datastreams/aiokafka.py` for DSM
Is that the correct approach to go about?

Reply:
Is there any existing simple integration that implements that pattern which I could use as a base?